File: swift-api/Sources/App/Models/Todo.swift
import Foundation
import Hummingbird

struct Todo: Sendable, Codable {
    let id: UUID
    let userId: UUID
    var title: String
    var order: Int?
    var completed: Bool
    var url: String?
    let createdAt: Date
    var updatedAt: Date

    init(
        id: UUID = UUID(),
        userId: UUID,
        title: String,
        order: Int? = nil,
        completed: Bool = false,
        url: String? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.userId = userId
        self.title = title
        self.order = order
        self.completed = completed
        self.url = url
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

struct TodoResponse: Codable, ResponseEncodable, Sendable {
    let id: UUID
    let title: String
    let order: Int?
    let completed: Bool
    let url: String
    let createdAt: Date
    let updatedAt: Date

    init(from todo: Todo, baseURL: String) {
        self.id = todo.id
        self.title = todo.title
        self.order = todo.order
        self.completed = todo.completed
        self.url = todo.url ?? "\(baseURL)/todos/\(todo.id)"
        self.createdAt = todo.createdAt
        self.updatedAt = todo.updatedAt
    }
}

File: go-api/internal/models/todo.go
package models

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type Todo struct {
	ID        uuid.UUID `json:"id"`
	UserID    uuid.UUID `json:"user_id"`
	Title     string    `json:"title"`
	Order     *int      `json:"order,omitempty"`
	Completed bool      `json:"completed"`
	URL       *string   `json:"url,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type TodoResponse struct {
	ID        uuid.UUID `json:"id"`
	Title     string    `json:"title"`
	Order     *int      `json:"order,omitempty"`
	Completed bool      `json:"completed"`
	URL       string    `json:"url"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

func (t *Todo) ToResponse(baseURL string) TodoResponse {
	url := fmt.Sprintf("%s/todos/%s", baseURL, t.ID.String())
	if t.URL != nil {
		url = *t.URL
	}

	return TodoResponse{
		ID:        t.ID,
		Title:     t.Title,
		Order:     t.Order,
		Completed: t.Completed,
		URL:       url,
		CreatedAt: t.CreatedAt,
		UpdatedAt: t.UpdatedAt,
	}
}

type CreateTodoRequest struct {
	Title string `json:"title" binding:"required"`
	Order *int   `json:"order,omitempty"`
}

type UpdateTodoRequest struct {
	Title     *string `json:"title,omitempty"`
	Order     *int    `json:"order,omitempty"`
	Completed *bool   `json:"completed,omitempty"`
}
